Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean

```diff
@@ -10,56 +10,177 @@ import Mathlib.GroupTheory.FreeGroup.Basic
-* `IsFreeGroup.of` - the canonical injection of `G`'s generators into `G`
-* `IsFreeGroup.lift` - the universal property of the free group
+* `FreeGroupBasis ι G` : a function from `ι` to `G` such that `G` is free over its image.
+  Equivalently, an isomorphism between `G` and `FreeGroup ι`.
+
+* `IsFreeGroup G` : a typeclass to indicate that `G` is free over some generators
+* `Generators G` : given a group satisfying `IsFreeGroup G`, some indexing type over
+  which `G` is free.
+* `IsFreeGroup.of` : the canonical injection of `G`'s generators into `G`
+* `IsFreeGroup.lift` : the universal property of the free group
 
 ## Main results
 
-* `IsFreeGroup.toFreeGroup` - any free group with generators `A` is equivalent to `FreeGroup A`.
-* `IsFreeGroup.unique_lift` - the universal property of a free group
-* `IsFreeGroup.ofUniqueLift` - constructing `IsFreeGroup` from the universal property
+* `FreeGroupBasis.isFreeGroup`: a group admitting a free group basis is free.
+* `IsFreeGroup.toFreeGroup`: any free group with generators `A` is equivalent to `FreeGroup A`.
+* `IsFreeGroup.unique_lift`: the universal property of a free group.
+* `FreeGroupBasis.ofUniqueLift`: a group satisfying the universal property of a free group admits
+ a free group basis.
 
 -/
 
 
 universe u
 
+open Function Set
+
+noncomputable section
+
+/-- A free group basis `FreeGroupBasis ι G` is a structure recording the isomorphism between a
+group `G` and the free group over `ι`. One may think of such a basis as a function from `ι` to `G`
+(which is registered through a `FunLike` instance) together with the fact that the morphism induced
+by this function from `FreeGroup ι` to `G` is an isomorphism. -/
+structure FreeGroupBasis (ι : Type*) (G : Type*) [Group G] where
+  /-- `FreeGroupBasis.ofRepr` constructs a basis given an equivalence with a free group. -/
+  ofRepr ::
+    /-- `repr` is the isomorphism between the group `G` and the free group generated by `ι`. -/
+    repr : G ≃* FreeGroup ι
+
+/-- A group is free if it admits a free group basis. In the definition, we require the basis to
+be in the same universe as `G`, although this property follows from the existence of a basis in
+any universe, see `FreeGroupBasis.isFreeGroup`. -/
+class IsFreeGroup (G : Type u) [Group G] : Prop where
+  nonempty_basis : ∃ (ι : Type u), Nonempty (FreeGroupBasis ι G)
+
+namespace FreeGroupBasis
+
+variable {ι ι' G H : Type*} [Group G] [Group H]
+
+/-- A free group basis for `G` over `ι` is associated to a map `ι → G` recording the images of
+the generators. -/
+instance funLike : FunLike (FreeGroupBasis ι G) ι (fun _ ↦ G) where
+  coe := fun b i ↦ b.repr.symm (FreeGroup.of i)
+  coe_injective' := by
+    rintro ⟨b⟩  ⟨b'⟩ hbb'
+    suffices H : (b.symm : FreeGroup ι →* G) = (b'.symm : FreeGroup ι →* G) by
+      have : b.symm = b'.symm := by ext x; exact FunLike.congr_fun H x
+      have : b.symm.symm = b'.symm.symm := by rw [this]
+      simpa using this
```

Comment:
This should surely follow in one tactic from an existing injectivity lemma or a `symm_eq_symm` lemma or something?

Reply:
That's what I thought, but I couldn't locate a convenient lemma for that. Hence the slightly convoluted proof...

Reply:
I think in general we have been avoiding adding the Equivalent versions of any ext lemmas; if you want them, you can get there by applying `MulEquiv.toMonoidHom_injective` or similar. Tagging *these* lemmas with `ext` may be reasonable.